Applications/Spire/Include/Spire/Ui/AssociativeValueModel.hpp

```diff
@@ -0,0 +1,242 @@
+#ifndef SPIRE_ASSOCIATIVE_VALUE_MODEL_HPP
+#define SPIRE_ASSOCIATIVE_VALUE_MODEL_HPP
+#include <functional>
+#include <boost/signals2/shared_connection_block.hpp>
+#include "Spire/Ui/Ui.hpp"
+#include "Spire/Ui/LocalValueModel.hpp"
+
+namespace Spire {
+namespace Details {
+  struct AssociatedBooleanModel {
+    std::shared_ptr<BooleanModel> m_model;
+    boost::signals2::scoped_connection m_connection;
+  };
+
+  template <typename T, typename U = boost::optional<T>>
+  struct OptionalHash {
+    std::size_t operator()(const U& value) const;
+  };
+}
+
+  /** 
+   * Associates BooleanModels with corresponding values, and ensures that
+   * only a single model has a value of true at one time.
+   */
+  template<typename T, typename U = std::hash<T>>
```

Comment:
Remove the ```typename U```.

Reply:
What would a solution like this look like? The problem I was having was how to properly call the right hash function when differentiating the optional< T > type from other types.

Reply:
You don't differentiate it. There's no need to differentiate based on hashing.

Reply:
But there's no hash struct for the boost::optional type, and I need to pass it as a template parameter for the unordered_map. As far as I know this hash implementation would then need some way to calculate the correct hash based on whether it's an optional or another type.

Reply:
Good point, look at something like this link to resolve a problem like this:

https://stackoverflow.com/questions/14813965/how-to-specialize-stdhash-for-type-from-other-library

This link goes into the more general solution:

https://en.cppreference.com/w/cpp/language/partial_specialization

This technique is reasonably advanced and very powerful, so feel free to take some time to familiarize yourself with it and then apply it to solve this issue.